### PR TITLE
loopback mode for profile_band, eta_phase estimator using IQ noise, fast data taking example scripts

### LINKED ISSUE #431: Support for loopback in `profile_band`
## Describe the problem
I swear I did this already, but I can't find it. We want to have `profile_band` work for loopback mode.

### PR DESCRIPTION
**Pull request by eyyoung24**
## Issue
This PR resolves #431, #443, 

## Description

- loopback mode for `profile_band`
- High bandwidth noise measurement function
- High bandwidth noise eta rotation
- Example scripts of the two above
- Convenience function for going from feedback_start/end_frac to feedback
- Fixes bug that was preventing us from configuring for arbitrary flux ramp rates if not using a timing system (see issue https://github.com/slaclab/pysmurf/issues/443).
- New functions:
-- `smurf_command.set_rf_iq_stream_enable` : sets enable bit which controls streaming RF I&Q data through `take_debug_data`.
-- `smurf_command.get_rf_iq_stream_enable` : gets enable bit which controls streaming RF I&Q data through `take_debug_data`.
-- `smurf_noise.take_noise_high_bandwidth` : takes 2.4 MHz fixed tone data on a single channel.
-- `smurf_tune._feedback_frac_to_feedback` : convenience function for computing the raw feedbackStart/feedbackEnd register values.
-- `smurf_tune._feedback_to_feedback_frac` : convenience function ; inverse of `smurf_tune._feedback_frac_to_feedback`.
-- `smurf_tune.calculate_eta_svd` : alternate single channel eta parameter estimator ; uses SVD on single channel I and Q data.
-- `smurf_tune.optimize_lms_delay` : attempts to find the best value for lmsDelay by measuring df vs lmsDelay for all active channels.
- `flux_mod2` - some blocks are 1 shorter. If there's a shorter block, this would crash the script. This fixes that bug. 

## Does this PR break any interface?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by eyyoung24** [comment]
@jmdewart I'm not actually sure I commented the flux ramp stuff properly. In particular, I'm confused by what's going on in `flux_ramp_rate_to_PV`

**Comment by swh76** [comment]
@eyyoung24 can you list the new functions with a brief description in the PR description that you're adding to client (not in your scratch and stuff like that)?

File: python/pysmurf/client/debug/smurf_noise.py [hunk 1]
```diff
@@ -1802,7 +1802,8 @@ def analyze_noise_vs_tone(self, tone, datafile, channel=None, band=None,
 
 
     def take_noise_high_bandwidth(self, band, channel, tone_power=10,
-        nsamp=2**25):
+        nsamp=2**25, nperseg=2**18, make_plot=True, show_plot=False,
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/debug/smurf_noise.py]: Docstring was added in a later commit.

File: python/pysmurf/client/tune/smurf_tune.py [hunk 2]
```diff
@@ -3175,7 +3175,8 @@ def check_lock(self, band, f_min=.015, f_max=.2, df_max=.03,
             lms_freq_hz=lms_freq_hz, reset_rate_khz=reset_rate_khz,
             feedback_start_frac=feedback_start_frac,
             feedback_end_frac=feedback_end_frac, lms_enable1=lms_enable1,
-            lms_enable2=lms_enable2, lms_enable3=lms_enable3)
+            lms_enable2=lms_enable2, lms_enable3=lms_enable3,
+            lms_gain=lms_gain)
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/tune/smurf_tune.py:3248]: Added to docstring.

File: python/pysmurf/client/tune/smurf_tune.py [hunk 3]
```diff
@@ -3104,7 +3104,8 @@ def get_fraction_full_scale(self, new_epics_root=None):
     def check_lock(self, band, f_min=.015, f_max=.2, df_max=.03,
             make_plot=False, flux_ramp=True, fraction_full_scale=None,
             lms_freq_hz=None, reset_rate_khz=None, feedback_start_frac=None,
-            feedback_end_frac=None, setup_flux_ramp=True, **kwargs):
+            feedback_end_frac=None, setup_flux_ramp=True, lms_enable1=None,
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/tune/smurf_tune.py:3156]: Added to docstring.

File: python/pysmurf/client/debug/smurf_noise.py [hunk 4]
```diff
@@ -1817,11 +1817,18 @@ def take_noise_high_bandwidth(self, band, channel, tone_power=10,
         freq = self.channel_to_freq(band, channel)  # resonance frequency
         channel_freq = self.get_channel_frequency_mhz(band) * 1.0E6 # sampling freq
 
+        if tone_power is None:
+            tone_power = self.get_amplitude_scale_channel(band, channel)
+            if tone_power == 0:
+                self.log("Tone amplitude is set to zero. Using default of 10")
+                tone_power = 10
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/debug/smurf_noise.py]: Changed from 10 to 12.

File: python/pysmurf/client/debug/smurf_noise.py [hunk 5]
```diff
@@ -1855,16 +1855,16 @@ def take_noise_high_bandwidth(self, band, channel, tone_power=None,
         if make_plot:
             fig, ax = plt.subplots(2, figsize=(6,5.5),
                 gridspec_kw={'height_ratios': [1, 2]})
-            t = np.arange(nperseg)/channel_freq * 1.0E6 # Time in us
+            t = np.arange(nperseg)/channel_freq * 1.0E3 # Time in ms
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/debug/smurf_noise.py]: Done.

File: python/pysmurf/client/util/smurf_util.py [hunk 6]
```diff
@@ -3884,7 +3884,7 @@ def get_gradient_descent_params(self, band):
         return ret
 
 
-    def set_fixed_tone(self,freq_mhz,drive,quiet=False):
+    def set_fixed_tone(self, freq_mhz, drive, write_log=False):
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/util/smurf_util.py:3867]: Ret->Returns, added write_log to the docstring.

File: python/pysmurf/client/tune/smurf_tune.py [hunk 7]
```diff
@@ -3773,9 +3780,67 @@ def last_tune(self):
     def optimize_lms_delay(self, band, lms_delays=None, reset_rate_khz=None,
         fraction_full_scale=None, nsamp=2**18, lms_gain=7, lms_freq_hz=None,
         meas_lms_freq=False, feedback_start_frac=.2, feedback_end_frac=.98,
-        meas_flux_ramp_amp=True, make_plot=False, show_plot=False,
+        meas_flux_ramp_amp=True, n_phi0=4, make_plot=False, show_plot=False,
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/tune/smurf_tune.py]: It's False after all commits, so I think meas_flux_ramp_amp must get set to False by default in a later commit.  Added missing docstring args.

File: python/pysmurf/client/tune/smurf_tune.py [hunk 8]
```diff
@@ -3769,6 +3769,52 @@ def last_tune(self):
         return np.sort(glob.glob(os.path.join(self.tune_dir,
                                               '*_tune.npy')))[-1]
 
+    @set_action()
+    def optimize_lms_delay(self, band, lms_delays=None, reset_rate_khz=None,
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/tune/smurf_tune.py:3998]: Here's the list, I think ;

`smurf_command` : `set_rf_iq_stream_enable`
`smurf_command` : `get_rf_iq_stream_enable`

`smurf_noise` : `take_noise_high_bandwidth`

`smurf_tune` : `_feedback_frac_to_feedback`
`smurf_tune` : `_feedback_to_feedback_frac`
`smurf_tune` : `calculate_eta_svd`
`smurf_tune` : `optimize_lms_delay`

[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/tune/smurf_tune.py:3998]: Added brief descriptions of each of these new routines to PR description.

File: python/pysmurf/client/tune/smurf_tune.py [hunk 9]
```diff
@@ -3095,8 +3095,9 @@ def flux_ramp_setup(self, reset_rate_khz, fraction_full_scale, df_range=.1,
             write_log=write_log)
         self.set_enable_ramp_trigger(EnableRampTrigger, new_epics_root=new_epics_root,
             write_log=write_log)
-        self.set_ramp_rate(reset_rate_khz, new_epics_root=new_epics_root,
-            write_log=write_log)
+        if self.get_ramp_start_mode() == 1:
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/tune/smurf_tune.py:3133]: Added a note about this to the PR description with a link to the issue.  Added a link to the code fix that fixed the issue on the issue comments.

[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/tune/smurf_tune.py:3137]: Added a comment.

File: python/pysmurf/client/tune/smurf_tune.py [hunk 10]
```diff
@@ -3158,13 +3159,16 @@ def check_lock(self, band, f_min=.015, f_max=.2, df_max=.03,
...
 
         if fraction_full_scale is None:
-            fraction_full_scale = self.fraction_full_scale
+            # fraction_full_scale = self.fraction_full_scale
+            fraction_full_scale = self.get_fraction_full_scale()
 
         if lms_freq_hz is None:
-            lms_freq_hz = self.lms_freq_hz[band]
+            # lms_freq_hz = self.lms_freq_hz[band]
+            lms_freq_hz = self.get_lms_freq_hz(band)
```
[the remark this answers is not in the document]
Comment (eyyoung24) [reply python/pysmurf/client/tune/smurf_tune.py:3231]: This is in `check_lock`. So this makes `check_lock` work as intended (was a bug before). We want to have `check_lock` use the existing value from `tracking_setup` by default.

File: python/pysmurf/client/tune/smurf_tune.py [hunk 11]
```diff
@@ -3000,9 +3000,13 @@ def set_fixed_flux_ramp_bias(self,fractionFullScale,debug=True,
     def flux_ramp_setup(self, reset_rate_khz, fraction_full_scale, df_range=.1,
             band=2, write_log=False, new_epics_root=None):
         """
-        Set flux ramp sawtooth rate and amplitude. If there are errors, check
-        that you are using an allowed reset rate! Not all rates are allowed.
-        Allowed rates: 1, 2, 3, 4, 5, 6, 8, 10, 12, 15 kHz
+        Set flux ramp sawtooth rate and amplitude. 
+
+        If you are using a timing system , Flux ramp reset rate must integer
```
[the remark this answers is not in the document]
Comment (swh76) [reply python/pysmurf/client/tune/smurf_tune.py]: Fixed.

**Pull request merged**